src/coreclr/jit/llvm.cpp

```diff
@@ -1643,47 +1405,298 @@ llvm::DILocation* createDebugFunctionAndDiLocation(struct DebugMetadata debugMet
+    strcpy(txtFileName, _outputFileName);
+    strcpy(txtFileName + strlen(_outputFileName) - 2, "txt");
+    llvm::raw_fd_ostream textOutputStream(txtFileName, ec);
+    _module->print(textOutputStream, (llvm::AssemblyAnnotationWriter*)NULL);
+    free(txtFileName);
+#endif //DEBUG
+    llvm::raw_fd_ostream OS(_outputFileName, ec);
+    llvm::WriteBitcodeToFile(*_module, OS);
+    delete _module;
+    //    Module.Verify(LLVMVerifierFailureAction.LLVMAbortProcessAction);
+}
+
+void Llvm::ConvertShadowStackLocalNode(GenTreeLclVarCommon* node)
+{
+    GenTreeLclVarCommon* lclVar = node->AsLclVarCommon();
+    LclVarDsc* varDsc = _compiler->lvaGetDesc(lclVar->GetLclNum());
+    genTreeOps oper = node->OperGet();
+    if (!canStoreLocalOnLlvmStack(varDsc))
+    {
+        // TODO-LLVM: if the offset == 0, just GT_STOREIND at the shadowStack
+        GenTreeIntCon* offset = _compiler->gtNewIconNode(varDsc->GetStackOffset(), TYP_I_IMPL);
+        GenTreeLclVar* shadowStackVar = _compiler->gtNewLclvNode(_shadowStackLclNum, TYP_I_IMPL);
+        GenTree* lclAddress = _compiler->gtNewOperNode(GT_ADD, TYP_I_IMPL, shadowStackVar, offset);
+
+        genTreeOps indirOper = GT_NONE;
+        GenTree* storedValue = nullptr;
+        switch (node->OperGet())
+        {
+            case GT_STORE_LCL_VAR:
+                indirOper = lclVar->TypeIs(TYP_STRUCT) ? GT_STORE_OBJ : GT_STOREIND;
+                storedValue = node->AsOp()->gtGetOp1();
+                break;
+            case GT_LCL_VAR:
+                indirOper = lclVar->TypeIs(TYP_STRUCT) ? GT_OBJ : GT_IND;
+                break;
+            default:
+                unreached();
+        }
+        node->ChangeOper(indirOper);
+        node->AsIndir()->SetAddr(lclAddress);
+        if (GenTree::OperIsStore(indirOper))
+        {
+            node->gtFlags |= GTF_IND_TGT_NOT_HEAP;
+            node->AsOp()->gtOp2 = storedValue;
+        }
+
+        CurrentRange().InsertBefore(node, offset);
+        CurrentRange().InsertAfter(offset, shadowStackVar);
+        CurrentRange().InsertAfter(shadowStackVar, lclAddress);
+    }
+}
+
+void Llvm::ConvertShadowStackLocals()
+{
+    _shadowStackLclNum = _compiler->lvaGrabTemp(true DEBUGARG("shadowstack"));
+    LclVarDsc* shadowStackVarDsc = _compiler->lvaGetDesc(_shadowStackLclNum);
+    shadowStackVarDsc->lvIsParam = 1;
+    shadowStackVarDsc->lvType = var_types::TYP_I_IMPL;
+
+    for (BasicBlock* _currentBlock : _compiler->Blocks())
+    {
+        _currentRange = &LIR::AsRange(_currentBlock);
+        for (GenTree* node : CurrentRange())
+        {
+            if (node->OperIs(GT_STORE_LCL_VAR, GT_LCL_VAR))
+            {
+                ConvertShadowStackLocalNode(node->AsLclVarCommon());
+            }
+            else if (node->IsCall())
+            {
+                GenTreeCall* callNode = node->AsCall();
+
+                if (callNode->gtCallType == CT_HELPER || callNode->gtReturnType == var_types::TYP_VOID)
```

Comment:
```suggestion
                if (callNode->IsHelperCall() || callNode->TypeIs(TYP_VOID))
```
`gtReturnType` is only really different for small typed calls.